core: add train sim v3

hhirtz opened the pull request:
Tracking issue: https://github.com/OpenRailAssociation/osrd/issues/13273
Previous PR: https://github.com/OpenRailAssociation/osrd/pull/13522

cc @eckter @Erashin @flomonster @woshilapin @Tristramg @emersion @Tguisnet @Castavo

just a work in progress for now

hhirtz committed:
core: add train sim v3

hhirtz committed:
core: add train sim v3

hhirtz committed:
core: account for stock length in MRSP and neutral zones

hhirtz committed:
core: account for stock length in MRSP and neutral zones

hhirtz committed:
core: snap train speed to speed constraints

this makes it so that when the rolling stock reaches a speed close to a
limit (either the current speed limit or one of the deceleration curve),
the integration step is fixed to "snap" the speed to the limit.

The threshold of closeness is 1e-4 (chosen arbitrarily)

I think this is needed because of floating point errors, but also
because deceleration curves are interpolated as segmented lines (instead
of a quadratic curve), although i haven't tested if changing how we
interpolate them fixes the problem.

Additionally, the time step may be decreased to improve correctness.

hhirtz committed:
core: handle stops

hhirtz committed:
core: snap train speed to speed constraints

this makes it so that when the rolling stock reaches a speed close to a
limit (either the current speed limit or one of the deceleration curve),
the integration step is fixed to "snap" the speed to the limit.

The threshold of closeness is 1e-4 (chosen arbitrarily)

I think this is needed because of floating point errors, but also
because deceleration curves are interpolated as segmented lines (instead
of a quadratic curve), although i haven't tested if changing how we
interpolate them fixes the problem.

Additionally, the time step may be decreased to improve correctness.

hhirtz committed:
core: handle stops

hhirtz committed:
core: account for the rolling stock's max speed

hhirtz committed:
core: Make a standalone simulation out of train-sim v3

i'd like to integrate v3 into `runStandaloneSimulation` instead of
replacing it, it's most likely less work


Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
core: Make a standalone simulation out of train-sim v3

i'd like to integrate v3 into `runStandaloneSimulation` instead of
replacing it, it's most likely less work


Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
core: account for the rolling stock's max speed

hhirtz commented:
Commit cb8302c151a0a004567b0c70d210e74dd49f1531 should work correctly, at least enough to show a (slightly off) speed diagram:

<img width="1005" height="458" alt="screen-20251104125411" src="https://github.com/user-attachments/assets/2d174600-3b38-4762-9f85-77b20abfe494" />

I had to do some snapping in 9ab5980302505716c1e760df93b423567c2b6a82 otherwise the train speed can jigsaw indefinitely like below. Probably a lot better than to have the train go slightly off-speed for 3 hours and arrive way early or late.

<img width="1055" height="461" alt="screen-20251103155828" src="https://github.com/user-attachments/assets/6d77fdf3-5263-4742-9ef0-c65af5cbde77" />

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/onetrain.kt:
```diff
@@ -0,0 +1,164 @@
+            mrsp.putLower(Range.closed(start, end), speed)
+        }
+    }
+    mrsp = mrsp.withStockLength(rollingStock.length)
```
It's not quite that simple: the rolling stock length only applies to *some* kinds of speed limits. Specifically, the speed limits caused by stops on blocking signals are not affected by this.

hhirtz replied:
aeba35f7a7a5fbad7572ca6db77f6dde708ff326 moved the call just after the speed limit properties, like what `computeMRSP` does i think.

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/onetrain.kt:
```diff
@@ -0,0 +1,164 @@
+    }
+    mrsp = mrsp.withStockLength(rollingStock.length)
+    for ((position, _) in stops) {
+        mrsp.put(Range.closed(position, position), 0.0)
```
Expressing a stop as a speed limit can be tricky, it's harder to control when and how the train stops and restarts. At some point that constraint needs to be lifted.

Not saying it can't work that way, just that it *might* be more complex further down. 

(the first simulation implementation did it that way and it added a lot of extra `if (...)` for all the edge cases)

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/onetrain.kt:
```diff
@@ -0,0 +1,164 @@
+    for ((stopPosition, stopDuration) in stops) {
+        while (!(stopPosition approxLowerThan position)) {
+            val s = step(ctx, instructions, timeStep, position, speed)
+            assert(s.positionDelta > 1e-6) { "le train c stopper.... ${s.positionDelta}" }
```
It's hard to tell exactly when this is an error case, for example if the previous step is one epsilon away from a stop, we might get epsilon-length steps.

Plus, the v3 is supposed to properly handle times and positions at the same time. We could imagine a step with 0 position delta and >0 time delta, for example while waiting at a stop. 

(Though this is definitely good enough for now, especially as infinite loops are more of an issue than errors)

hhirtz replied:
I'll change the value of `1e-6` to not be hardcoded here, but the loop condition already ensures the train is at least `1e-6` meters away from the stop.

Currently below the loop there's a time skip to after the stop. I don't think it makes sense to call `step` while waiting at a stop, since we know the train isn't going to move. When we add pantograph logic (and other state like battery level?), those can be updated independently.

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/onetrain.kt:
```diff
@@ -0,0 +1,164 @@
+                    val position = offset.meters
+                    var res = simplifiedPoints.binarySearchBy(position) { point -> point.position }
+                    if (res < 0) {
+                        // TODO offset by one to have the point after arival instead of before?
```
Here we want the arrival time and not departure time. But that distinction is indeed often important

hhirtz replied:
fixed in 5e434ccc1a26ccdc86cd8d04ad5f02a757ed0b64

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/range.kt:
````diff
@@ -0,0 +1,89 @@
+ * assert(map.get(0.0) == 69.0)
+ * ```
+ */
+internal fun RangeMap<Double, Double>.putLower(range: Range<Double>, value: Double) {
````
There's a `RangeMap` function that might be relevant here:

```
  void merge(
      Range<K> range,
      @CheckForNull V value,
      BiFunction<? super V, ? super @Nullable V, ? extends @Nullable V> remappingFunction);
```

It sets the range to the value, calling the function on intersections.

hhirtz replied:
thanks for the pointer, fixed in 2ba2d602e305944fde7de6e33d8376d97455b5f5

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/tsim.kt:
```diff
@@ -0,0 +1,488 @@
+
+typealias MeterArray = DoubleArray
+
+typealias MeterPerSecondArray = DoubleArray
```
We already have a `Distance` type [there](https://github.com/OpenRailAssociation/osrd/blob/dev/core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/units/Distance.kt#L31). 

It's an inline value type, internally it's just a `Long`. But it has better type safety and more abstraction on the underlying unit.

There's also a `DistanceArray` type to avoid boxing, but that part isn't used much. Maybe we should, if we can make it implement some `List<Distance>` interface first for convenience.

There's also `Offset<T>` that can be used to type the distance reference point. In the train sim, I'd expect most distances to be `Offset<TrainPath>`. 

And there's similar units for speeds and times. They're not used as much because we didn't want to convert the simulation module, but now would be a good time.

**Note**: these types are not floating point. This *might* be tricky to deal with when there's no lower bound on step lengths. We need to be very careful about rounding in these cases.

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/tsim.kt:
```diff
@@ -0,0 +1,488 @@
+        val speedLimit = entry.value
+
+        // The speed limit is enforced as long as part of the rolling stock is behind the reset sign
+        // TODO what's the name of the sign that reset a speed limit?
```
Well I don't know :sweat_smile:

But we don't generally need to deal with this in core, speed limits are directly described as ranges in the infra description json.

flomonster replied:
At sncf the sign looks like this (R is for "reprise de vitesse").

![Panneau reprise](https://upload.wikimedia.org/wikipedia/commons/thumb/3/38/Tableau_R.svg/250px-Tableau_R.svg.png)

A sign is not always there to indicate the end of a speed limit; sometimes it is just information passed on verbally between drivers.

Like @eckter said, we try to avoid using SNCF specific vocabulary in core (when possible).

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/tsim.kt:
```diff
@@ -0,0 +1,488 @@
+ * This class represents a list of 2D points `(xs[i],ys[i])` connected by straight lines. [xs] and
+ * [ys] must have the same size. The elements of [xs] are expected to be strictly increasing.
+ */
+internal class Curve(val xs: DoubleArray, val ys: DoubleArray) {
```
After learning from our mistakes with `Envelope`: there's a good chance we'll need a single interface that can query "arrival time at" and "departure time from", which needs to account for stops. This object here is very close to what we had with `Envelope`, and that would mean wrapping it into a different object to add stop data. Which would result in the current `EnvelopeStopWrapper`.

If we go that route, similar to what we already have, I'd prefer if we keep the `Envelope` type mostly as it is. It's not trivial code, it works fine, has good tooling, and it's been properly tested and tweaked. (Not necessarily the "overlay builders" and such, just the base type itself)

But I *think* I'd have a slight preference for a new types that directly include stop data right from the start.

eckter replied:
Note: if we're truly aiming for dynamic simulations, then the stop exact duration is unknown up to the actual departure. That's often an issue when trying to unify it all into a single interface. 

We use workarounds to handle that case in STDCM, with stop duration that are set to "infinity" until it gets replaced later on.

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/tsim.kt:
```diff
@@ -0,0 +1,488 @@
+     */
+    private val decelerationCurves = mutableMapOf<DecelerationTarget, Curve>()
+
+    internal fun decelerationCurve(target: DecelerationTarget, dt: Seconds): Curve {
```
Note: deceleration curves will eventually need to be part of the exposed API

eckter commented on core/src/main/kotlin/fr/sncf/osrd/tsim/tsim.kt:
```diff
@@ -0,0 +1,488 @@
+ * [speed] for a given time [dt].
+ *
+ * The context [ctx] may be reused between calls to improve performance.
+ */
```
I appreciate the nicely formatted docstrings :+1: I need to do that more often

eckter reviewed (commented):
Really nice work :+1: I like where this is going

I tried to keep to "macroscopic" comments and decisions that we need to make early. These comments are of course open to discussion.

hhirtz committed:
Fix pathItemTimes off by one

hhirtz committed:
Only apply stock length on speed limit properties

like what's computeMRSP is doing

hhirtz committed:
implement RangeMap.putLower with RangeMap.merge

hhirtz committed:
Only apply stock length on speed limit properties

like what's computeMRSP is doing

hhirtz committed:
implement RangeMap.putLower with RangeMap.merge

hhirtz committed:
Fix pathItemTimes off by one

hhirtz committed:
core: rewrite withStockLength with putLower and remove TODO

the name of the sign is infrastructure manager specific

hhirtz committed:
core: rewrite withStockLength with putLower and remove TODO

the name of the sign is infrastructure manager specific

hhirtz committed:
rename Instructions.mrsp to speedLimits

Since the field isn't used just for the mrsp

hhirtz committed:
rename Instructions.mrsp to speedLimits

Since the field isn't used just for the mrsp

hhirtz committed:
core: rework algorithmic structure

hhirtz committed:
core: rework algorithmic structure

hhirtz committed:
core: start simulation at position 0.0

hhirtz committed:
core: fix truncateStep

hhirtz committed:
là je fais des bails sur tsim

hhirtz committed:
core: use quad interpolation and rework reactToSpeedConstraint

quad interpolation helps because (i believe) the deceleration curves
are like 2nd order polynomials

hhirtz committed:
core: add arrival to stops even if arrival stop time is 0

hhirtz committed:
core: add mrsp to report to see mrsp on the STD

hhirtz committed:
core: start simulation at position 0.0

hhirtz committed:
core: add mrsp to report to see mrsp on the STD

hhirtz committed:
core: use quad interpolation and rework reactToSpeedConstraint

quad interpolation helps because (i believe) the deceleration curves
are like 2nd order polynomials

hhirtz committed:
core: also sort reactions end speed

hhirtz committed:
core: fix truncateStep

hhirtz committed:
core: add arrival to stops even if arrival stop time is 0

hhirtz committed:
là je fais des bails sur tsim

hhirtz committed:
core: trim report to the path length

since because of simulation errors the train can go overboard

hhirtz committed:
core: also sort reactions end speed

hhirtz committed:
core: trim report to the path length

since because of simulation errors the train can go overboard

hhirtz commented:
Results of the kotlin native investigation in this other PR: #13998

hhirtz committed:
several things...

- merge the lone brake into the list of reactions
- aiming to be general on constraints
- also compare reactions by timeDeltas
- add neutral zones
- fix maintain
- fix braking step truncation
- remove unused stuff
- remove redundant arguments in reactToSpeedConstraint
- snap to the curve on braking
- random asserts

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
several things...

- merge the lone brake into the list of reactions
- aiming to be general on constraints
- also compare reactions by timeDeltas
- add neutral zones
- fix maintain
- fix braking step truncation
- remove unused stuff
- remove redundant arguments in reactToSpeedConstraint
- snap to the curve on braking
- random asserts

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
the naive step always accelerates

since the currentSpeedLimit is handled as a reaction

hhirtz committed:
LTV example: remove MRSP from the space-time diagram

hhirtz committed:
the naive step always accelerates

since the currentSpeedLimit is handled as a reaction

hhirtz committed:
don't maintain speed if the speed limit is 0.0

hhirtz committed:
LTV example for InA

hhirtz committed:
LTV example for InA

hhirtz committed:
fix typo s/xs/ys/ and don't truncate if the endLimit is higher than the startSpeed

hhirtz committed:
LTV example: remove MRSP from the space-time diagram

hhirtz committed:
fix typo s/xs/ys/ and don't truncate if the endLimit is higher than the startSpeed

hhirtz committed:
don't maintain speed if the speed limit is 0.0

hhirtz committed:
When on the curve, return the next curve point

to do that we intersect the step and the curve by intersecting the step
and all curve segments. the curve is thus interpolated linearly (instead
of quadratically before)

this makes it so that the intersection point found is the same point
we'd get by doing constraint.lerp and it allows us to snap to the curve
correctly

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
When on the curve, return the next curve point

to do that we intersect the step and the curve by intersecting the step
and all curve segments. the curve is thus interpolated linearly (instead
of quadratically before)

this makes it so that the intersection point found is the same point
we'd get by doing constraint.lerp and it allows us to snap to the curve
correctly

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
lmao


Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
lmao


Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
random name changes

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
random name changes

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz commented:
I took off from before the double->long changes, sorry. i'll rebase and sort out the bugs later

commit 4aa87c911a1055f3251f8c79e1b02378756a3359 is for an example of LTV like: "after 12pm, max speed is limited to 200km/h." here `2934.0` is hardcoded because the train starts at 11:11:26am, but it can be computed if needed i guess.

before:
<img width="1292" height="454" alt="screen-20251126160852" src="https://github.com/user-attachments/assets/43de5dfb-6d21-4b89-85d8-a9295c9e293f" />


after:
<img width="1292" height="454" alt="screen-20251126160702" src="https://github.com/user-attachments/assets/d372822a-3a7d-4e65-9edd-6c37f40c5277" />

Handling the LTV like this means we don't anticipate the 200km/h speed limit, so the train ends up going too fast for a moment, but it shows how speed limits can change during the simulation.

hhirtz committed:
use Long instead of Double


Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
use Long instead of Double


Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
bon allez je vais push comme un ok

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
bon allez je vais push comme un ok

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

MartinBourbier committed:
core: merci les parenthèses

Signed-off-by: Martin Bourbier <mbourbier28@gmail.com>

MartinBourbier committed:
core: merci les parenthèses

Signed-off-by: Martin Bourbier <mbourbier28@gmail.com>

hhirtz committed:
move from divX to signalinglong

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
move from divX to signalinglong

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
wip new API

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
wip new API

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
firstAfterStrict returns null if out of bounds

forces caller to handle the out of bounds case

hhirtz committed:
spotlessApply

hhirtz committed:
firstAfterStrict returns null if out of bounds

forces caller to handle the out of bounds case

hhirtz committed:
spotlessApply

hhirtz committed:
session coworking

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
session coworking

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
questions for later

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
questions for later

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

MartinBourbier committed:
feat: implement a function to merge deceleration curves

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

MartinBourbier committed:
feat: implement a function to merge deceleration curves

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

MartinBourbier committed:
wip

Signed-off-by: Martin Bourbier <mbourbier28@gmail.com>

MartinBourbier committed:
wip

Signed-off-by: Martin Bourbier <mbourbier28@gmail.com>

hhirtz committed:
peer

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
peer

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add typed units

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add typed units

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
squelette de TrainState.merge

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
squelette de TrainState.merge

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add asserts in step

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add asserts in step

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
make lower/raisePantographTime optional

because it was optional in RollingStock

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
make lower/raisePantographTime optional

because it was optional in RollingStock

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
impl toString for most classes

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
impl toString for most classes

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
nullable SpeedConstraint.speedCurve in order to not apply the constraint

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
nullable SpeedConstraint.speedCurve in order to not apply the constraint

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add requires to nextStates

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add requires to nextStates

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
truncate step when the next point on the curve is further than context.timeStep

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
truncate step when the next point on the curve is further than context.timeStep

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
solve TODO return naive step

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
solve TODO return naive step

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
truncate step to the length of the path if it goes over

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
truncate step to the length of the path if it goes over

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
random commit just nothing of note here

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
random commit just nothing of note here

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
take into account pantograph position for acceleration

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
take into account pantograph position for acceleration

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
drop 2nd call to enactDecision

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
drop 2nd call to enactDecision

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
set pantograph state to Up by default

this means the pantograph state from constraints that do not care about
it is given lowest priority when merging

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
set pantograph state to Up by default

this means the pantograph state from constraints that do not care about
it is given lowest priority when merging

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
change PantographState to store the position as a double in [0, 1]

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
change PantographState to store the position as a double in [0, 1]

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
random spotless apply

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
random spotless apply

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
remove enactDecision's unused maxDelta parameter

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
remove enactDecision's unused maxDelta parameter

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
handle thermal trains

if raisePantographTime is null, assume the train is fully thermal and
can accelerate even if pantographState isn't UP

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
handle thermal trains

if raisePantographTime is null, assume the train is fully thermal and
can accelerate even if pantographState isn't UP

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
relax requires on truncate

these would otherwise trigger when the truncated step ends up having the
same time or position as the oldState

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
relax requires on truncate

these would otherwise trigger when the truncated step ends up having the
same time or position as the oldState

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
update old reference to signalPosition in NeutralSection's docs

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
update old reference to signalPosition in NeutralSection's docs

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
SpeedConstraint: handle cases where the rolling stock is close to the end

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
SpeedConstraint: handle cases where the rolling stock is close to the end

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
NeutralSection: always try to accelerate in the neutral section

otherwise thermal trains cannot accelerate during neutral sections

in the future, it would be better to handle train modes directly in
trainsim.kt

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
NeutralSection: always try to accelerate in the neutral section

otherwise thermal trains cannot accelerate during neutral sections

in the future, it would be better to handle train modes directly in
trainsim.kt

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
NeutralSection: avoid truncating steps too much

if the truncated steps do not advance the time (happens when
currentState is close to one of the borders of the section), don't
truncate

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
NeutralSection: avoid truncating steps too much

if the truncated steps do not advance the time (happens when
currentState is close to one of the borders of the section), don't
truncate

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Simulation: rename ctx to context

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Simulation: rename ctx to context

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
SpeedConstraint: fix false nulls in enactDecision before start

previously enactDecision would wrongly return null when
currentState.position<start because of the speedCurve check. Thus the
constraint wouldn't be respected before start.

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
SpeedConstraint: fix false nulls in enactDecision before start

previously enactDecision would wrongly return null when
currentState.position<start because of the speedCurve check. Thus the
constraint wouldn't be respected before start.

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
SimpleRollingStock: add missing raisePantographTime

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
SimpleRollingStock: add missing raisePantographTime

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
make RangeMap.putLower generic over value type

used in tests to avoid the .micrometersPerSecond boilerplate

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
make RangeMap.putLower generic over value type

used in tests to avoid the .micrometersPerSecond boilerplate

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
trainsim: remove references to old tsim

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
trainsim: remove references to old tsim

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
RangeTest: move tests over to the new trainsim module

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
tsim: remove

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
RangeTest: move tests over to the new trainsim module

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
tsim: remove

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
trainsim, Simulation: allow several states from enactDecision

this allows using two curves for ShortSlipStop

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
trainsim, Simulation: allow several states from enactDecision

this allows using two curves for ShortSlipStop

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add Updatable and stops

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add Updatable and stops

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
step: use sequences

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
step: use sequences

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
move UnitExt to PreciseUnits

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
move UnitExt to PreciseUnits

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
TrainState: stricter requires around truncating functions

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
TrainState: stricter requires around truncating functions

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
step: require train is not going above maxSpeed

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
step: require train is not going above maxSpeed

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
SpeedConstraint: fix false negatives

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
SpeedConstraint: fix false negatives

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Simulation: fix driverBehaviour argument type

This specific function mimics the current standalone simulation
function, so the new, trainsim-specific Driver type mustn't be used

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Simulation: fix driverBehaviour argument type

This specific function mimics the current standalone simulation
function, so the new, trainsim-specific Driver type mustn't be used

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
core: add PhysicsRollingStock.isThermal

the new train simulation needs to know whether a train can still
accelerate when the pantographs (if it has pantographs) are not up.

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
core: add PhysicsRollingStock.isThermal

the new train simulation needs to know whether a train can still
accelerate when the pantographs (if it has pantographs) are not up.

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
TrainState: use PhysicsRollingStock.isThermal

instead of the raisePantographTime hack

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
TrainState: use PhysicsRollingStock.isThermal

instead of the raisePantographTime hack

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
random spotlessApply diffs jj absorb doesn't want to process

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
random spotlessApply diffs jj absorb doesn't want to process

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
move SpeedConstraint to a separate file and split its enactDecision

it can be used for stops

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
move SpeedConstraint to a separate file and split its enactDecision

it can be used for stops

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Constraint: remove doesApply and add defaults for all methods

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Constraint: remove doesApply and add defaults for all methods

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
fix 0-length stops

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
fix 0-length stops

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Curve: add toString

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Curve: add toString

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Add TrainState.zero

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Add TrainState.zero

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move TrainState.toEnvelopePoint to an extension function

It won't be moved to osrd-mp since it's core-specific

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move TrainState.toEnvelopePoint to an extension function

It won't be moved to osrd-mp since it's core-specific

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
move SignalingLong to osrd-mp

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
move SignalingLong to osrd-mp

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move PreciseUnits to osrd-mp

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move PreciseUnits to osrd-mp

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move Curve and CurveTest to osrd-mp

Needed to also implement LongArray.binarySearch because no binary-search
function is defined on arrays in common code.........

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move Curve and CurveTest to osrd-mp

Needed to also implement LongArray.binarySearch because no binary-search
function is defined on arrays in common code.........

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
update Simulation with the OffsetRangeMap change

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
update Simulation with the OffsetRangeMap change

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move SpeedConstraint to osrd-mp

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move SpeedConstraint to osrd-mp

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move trainsim.kt to osrd-mp

Needed to implement float padding because there's no way to do it in
common code.

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Move trainsim.kt to osrd-mp

Needed to implement float padding because there's no way to do it in
common code.

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
fix RangeTest.testWithStockLength

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
fix RangeTest.testWithStockLength

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
WIP compute provisional envelope

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
WIP compute provisional envelope

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
take initialSpeed into account

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
take initialSpeed into account

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add Updatable.reset

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add Updatable.reset

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
peer to peer

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
peer to peer

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add todos about interpolating pantograph positions

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add todos about interpolating pantograph positions

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Set stop condition for vMaxFactor search to the timeStep

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Set stop condition for vMaxFactor search to the timeStep

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Start with vMaxFactorMin==0

Even with this, vMaxFactor will always be strictly greater than zero, so
there is no more issue than starting with 0.1

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Start with vMaxFactorMin==0

Even with this, vMaxFactor will always be strictly greater than zero, so
there is no more issue than starting with 0.1

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

MartinBourbier committed:
fix: take vMaxFactor into account in speed limited zones

Signed-off-by: Martin BOURBIER <mbourbier28@gmail.com>

MartinBourbier committed:
fix: take vMaxFactor into account in speed limited zones

Signed-off-by: Martin BOURBIER <mbourbier28@gmail.com>

hhirtz committed:
Simulation: don't add SpeedLimitedZones where end>start

the mrsp building creates closed ranges with 1 element (`[a,a]`)

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Simulation: don't add SpeedLimitedZones where end>start

the mrsp building creates closed ranges with 1 element (`[a,a]`)

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
WIP debug interface

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Add a tracer to log simulation states

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
debug interface in javascript

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
debug interface in javascript

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: put constraint index in name to help match it

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: put constraint index in name to help match it

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: cache data bounds

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: cache data bounds

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: faster bound compute and split canvas in two

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: clear old data when loading new file to avoid oom

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: show constraint names

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Stop: show initialDuration instead of duration in toString

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
add speed curves to the trace logs

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: show position and speed on hovered point

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: draw constraint names above orange points

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
don't extend deceleration curves to position zero

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
inspector: fix too-much-args RangeErrors

With large arrays `Math.max(...this.ys)` would fail. This patch thus
implements min and max by hand.

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz commented:
Ok noting this here to help resume work on this later: there are several problems with speed constraints and constraint merges... here's an example simulation:

blue curves are "runs", the highest one is the max-effort run, the others are runs for computing the vMaxFactor using a binary search (actually the margin is set at "0%" here so there's no point making this search but anyway)

green curve is the relief

<img width="1287" height="751" alt="screen-20260423180141" src="https://github.com/user-attachments/assets/61a0e0e0-ae16-426f-b90f-e8515b51001c" />

The speed constraints make some jank at the end that makes the train reach the stop with non-zero speed. Zooming on the jank:
 
<img width="1287" height="751" alt="screen-20260423175938" src="https://github.com/user-attachments/assets/6bee7793-d7ab-42b3-a60c-a58c792ca59c" />

There's a speed constraint of 44.444444444444444 m/s constantly on this portion of the path that decides to make the train maintain its speed on the whole step, and there's a stop constraint. I think the stop constraint's deceleration curve starts at the orange point because it decides on a train state with the same speed but truncates the step. However for some reason the next step after that (the little blue point on the right of the big blue point) isn't the stop's orange point? so there's an issue with the decision merge.

After that
<img width="1287" height="751" alt="screen-20260423181802" src="https://github.com/user-attachments/assets/23d20fb2-1179-42d4-931f-d59bc67e9242" />

there's a point where the simulation returns very little steps. This is what we usually call the @Castavo problem: the stop constraint decelerates on the full step, and the speed constraint accelerates, but since the train is really close to the speed limit, the decision from the speed constraint has a really small delta time. During the merge, we take the deceleration of the stop decision, with the time of the speed constraint, so we end with very small steps.

Then the stop constraint sharply decelerates for some reason, which unlocks the situation:
<img width="1287" height="751" alt="screen-20260423182442" src="https://github.com/user-attachments/assets/a8667eaf-c67c-4e8f-a1a3-138ca18cae2f" />

Then we start with the castavo problem again, but this time we can see it more clearly (the little blue dot is on the orange line, meaning we take the stop deceleration with a smaller time step for some reason):
<img width="1287" height="751" alt="screen-20260423182534" src="https://github.com/user-attachments/assets/efdfbd6d-8ab0-4dd0-aa90-2a4ca90b1d61" />

Then at some point the stop constraint refuses to apply?? so the train accelerates:
<img width="1287" height="751" alt="screen-20260423182710" src="https://github.com/user-attachments/assets/9f52b3c9-8c4c-4fe4-8286-6a34189df82c" />

And now we follow the stop constraint correctly (although at this point the train goes too fast to stop in time):
<img width="1287" height="751" alt="screen-20260423182824" src="https://github.com/user-attachments/assets/150f2a14-1de5-467b-acf6-e952eee055d7" />


Here's the log file to open it in the inspector:
[sim.log](https://github.com/user-attachments/files/27018405/sim.log)

hhirtz committed:
make Updatable.reset mandatory

since an Updatable constraint probably needs to reset state anyways,
this change makes sure we don't forget to implement it

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
to fixup: adapt runSimulation to changes introduced in ae8d0c665eb90fd6b2c1d2d9a118d4e56fa764e7

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Stop: make duration and stopCurve private

avoid usage by others since those fields are not meant to be used outside

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
trainsim: name constraints in decision safeguards

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
trainsim: use the max dt for state merging

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

hhirtz committed:
Simulation: don't do the binary search when there's no margin

Signed-off-by: Hubert Hirtz <hubert@hirtz.pm>

MartinBourbier committed:
traisin: setup multitrain simulation

Signed-off-by: Martin Bourbier <mbourbier28@gmail.com>

MartinBourbier committed:
trainsim: basic intersection regulator

Signed-off-by: Martin Bourbier <mbourbier28@gmail.com>

hhirtz commented:
Split the PR in several branches @MartinBourbier 

- https://github.com/OpenRailAssociation/osrd/pull/18194
- https://github.com/OpenRailAssociation/osrd/pull/18196
- https://github.com/OpenRailAssociation/osrd/pull/18198

it doesn't look like you can create several PR stacks targeting the same PR.. so only the multi train one is in the UI :shrug:
